### DISCUSSION TITLE
Question about Cuda

### DISCUSSION THREAD

**Discussion by Nem-z** [vladmandic/sdnext#111] [Q&A] [answered]
Hey,

I finally get the webui to launch on Windows with the last commit (thank you for that).

Now I have a question:
I see in the `System Info` tab that my Cuda version is 11.8:
![SysInfo](https://user-images.githubusercontent.com/63961359/231852639-6374bbb2-05cd-4eed-9671-2fe5b0d17d5e.png) 
When running `nvidia-smi` it shows I have Cuda 12.1 installed:
![NvidiaSMI](https://user-images.githubusercontent.com/63961359/231852652-2589f167-c724-4134-85b8-182b0593a948.png)
Why the Cuda version differs? Do my performances may be lower because of that? Because if I remember correctly when I benchmarked with A1111 I got ~40 it/s at max and in this version I hit ~30 it/s.
I tried to re-installing Cuda toolkit as stated in #107 discussion.

**Comment by vladmandic** [comment, marked as answer]
your system cuda libs are 12.1, but torch is compiled against 11.8, so it reports that.
normally mixing major versions is a no-no, but this is a known good combo, so no issues.  

and yeah, rtx4090 should give you >40it/s, but only at higher batch sizes as it cannot get saturated enough otherwise. 30it/s is about as expected at batch size 1.

**Reply by Nem-z** [reply]
Thank you for the quick answer.
As for my performance issue I re-launched a test on A1111 and it gave me ~30it/s this time around. So I must find where my problem is but it seems it's not on the WebUI side.

**Reply by rbaechtold** [reply]
Good to know, I was asking myself the same. I only get around 25it/s with my RTX4090 (at different batch sizes) so I guess I have to keep digging for a solution too.

**Reply by gad-ass** [reply]
Woah, I have same issue, on 4090 I have less then 20it/s.
Was testing on different models & sampling methods.
![chrome_cHbnnEudml](https://user-images.githubusercontent.com/32915902/233794263-25f685b7-bb46-4f95-82b4-d448a3deab47.png)

**Reply by AlUlkesh** [reply]
@gadaszewski 
To better be able to compare speeds, try the benchmark from the `System Info` tab with the stock 1.5 model.

On my 4090 I get:

```
Batchsize 1: 24.66
Batchsize 2: 35.05
Batchsize 4: 40.58
```

**Comment by ExecuteAI** [comment]
Can anyone help me understand this? I have uninstalled all and started over from Nvidia drivers to Nvidia tool kit 11. And 12.1. Anaconda new install, Vlad Diffusion 3 times. Why will it not find my CUDA I am on a 2080 Ti and it worked with the old Stable Diffusion. 
![image](https://user-images.githubusercontent.com/79231405/234088276-e670c144-e754-4881-8792-52523715c635.png)

**Reply by ExecuteAI** [reply]
I reinstalled Torch many times, too!

**Reply by ExecuteAI** [reply]
And when I try to create with a Prompt saying Car I get this. 
![image](https://user-images.githubusercontent.com/79231405/234088791-622b10dc-bd3e-4e64-900e-f1c7e08aeea7.png)

**Reply by ExecuteAI** [reply]
Where do I fix this, it did not work in the launch.py file 
RuntimeError: "LayerNormKernelImpl" not implemented for 'Half'

**Reply by vladmandic** [reply]
what happens if you try to run `nvidia-smi` from command line? its installed by nvidia display drivers and if you don't have it, update display drivers and check path settings. that utility is used to detect nvidia gpu. and you did not even say is this linux or windows. plus yelling in 4 updates in 5 minutes does not help at all!!!

**Reply by ExecuteAI** [reply]
![image](https://user-images.githubusercontent.com/79231405/234090865-5ed04638-d756-4dec-b3df-fb86f6a9d92c.png)

**Reply by ExecuteAI** [reply]
![image](https://user-images.githubusercontent.com/79231405/234091223-d7797b86-aa24-41bb-9b93-99a99881593b.png)

**Reply by vladmandic** [reply]
you don't even need cuda installed. nvidia-smi is installed by display driver and on windows its typically in `Windows\System32`. its its there, but that folder is not in path - add it to system path. and in general, if that folder is not in path, then you have bigger problems with your windows. and if nvidia-smi is not there, well, like i said, update device drivers!

**Reply by ExecuteAI** [reply]
And Pytorch is also installed. 
![image](https://user-images.githubusercontent.com/79231405/234092620-636694bb-a31a-4c9c-b5c2-edd5503cf7e8.png)

**Reply by ExecuteAI** [reply]
I still get this RuntimeError: "LayerNormKernelImpl" not implemented for 'Half'

**Reply by ExecuteAI** [reply]
I tried this https://stackoverflow.com/questions/75641074/i-run-stable-diffusion-its-wrong-runtimeerror-layernormkernelimpl-not-implem

But that does not work.

**Reply by ExecuteAI** [reply]
This i what I get when i use this thead they mentioned. 
![image](https://user-images.githubusercontent.com/79231405/234097860-278af984-c203-4aa3-a936-8f472c36f963.png)

**Reply by ExecuteAI** [reply]
![image](https://user-images.githubusercontent.com/79231405/234098249-14c6a89b-7a80-418b-a85a-48606a76d8be.png)

**Reply by vladmandic** [reply]
yeah, sure lets try something from a random post. first, that is not relevant to you - your goal is to get your gpu detected to start with, not to run with cpu. second, that param is gone. doesn't exist. please don't post irrelevant info as its not helping. each time i get notification i spend less time actually improving the product. i've provided information what is wrong with your system, so work on that. i really don't want to start ignoring notifications, but i might have to.

**Reply by ExecuteAI** [reply]
I'm uncertain if I understand your answer. I have spent all day on this. And I really think you have a great product here. But I'm not a developer. I'm a layman with some insights in code, nothing more. 
I do not understand why I get this error. And there are no answers online as fare as I can see. SRY for asking in this forum.

**Reply by vladmandic** [reply]
this is not a developer level stuff, this is windows user stuff, not even advanced user. make sure that nvidia-smi is present and available in path. as simple as that.

**Reply by ExecuteAI** [reply]
That is fixed. It worked. Thanks! *

I'm asking about this. --precision full --no-half
Where do I set it in webut.bat is not editable? 
It is an error I get.  It was my secondary question in the other post. 

![image](https://user-images.githubusercontent.com/79231405/234101451-f9e426af-a4ab-4e4c-a5bb-663d54c6580e.png)

**Reply by ExecuteAI** [reply]
RuntimeError: "LayerNormKernelImpl" not implemented for 'Half'

**Reply by ExecuteAI** [reply]
![image](https://user-images.githubusercontent.com/79231405/234102684-bb3ea1a9-1b5c-45df-9a12-b90068493c63.png)

**Reply by ExecuteAI** [reply]
I set in this command because I think I need all the others in this bat file. 
Then I get this. 
![image](https://user-images.githubusercontent.com/79231405/234102861-3ff7a78c-c74f-42cf-9393-d9bdc8ebf7f3.png)

**Reply by ExecuteAI** [reply]
Then I try here. But I get the same error.
![image](https://user-images.githubusercontent.com/79231405/234103537-b17ef8b7-b7cb-41ac-a195-e6cd04391c38.png)

**Reply by ExecuteAI** [reply]
I removed the first command line, but it still does not work.

**Reply by ExecuteAI** [reply]
And suddenly the CUDA does not respond again? 
![image](https://user-images.githubusercontent.com/79231405/234104649-5f5cb273-e831-4d39-ae44-39e45b39d3f5.png)

**Reply by ExecuteAI** [reply]
It worked when I made the path. Now it is not available. LOL

**Reply by vladmandic** [reply]
> I'm asking about this. --precision full --no-half

in UI settings -> CUDA params.

**Reply by ExecuteAI** [reply]
It finally worked after the fifth install. 
You write that the CUDA has to be installed, but you made me make a path instead. 
Look under your installation instructions. The path worked.

**Comment by ExecuteAI** [comment]


**Reply by ExecuteAI** [reply]
or does it have to be under system Variable?

**Reply by vladmandic** [reply]
system variable PATH always exists on windows. you should just add C:\Windows\System32 to it, not create new variable.

**Reply by ExecuteAI** [reply]
Ok I did. Thanks ill try now.
